File: Leetcode/Python/_1502.py
class Solution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        arr = sorted(arr)
        dif = arr[1] - arr[0]
        for i in range(len(arr)-1):
            if arr[i+1] - arr[i] != dif:
                return False
        return True
      
class Solution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        def sortInplace(array):
            for j in range(1, len(array)):
                key = array[j]
                i = j - 1
                while i > -1 and key>array[i]:
                    array[i+1] = array[i]
                    i -= 1
                array[i+1] = key
            return array
        arr = sortInplace(arr)
        dif = arr[1] - arr[0]
        for i in range(len(arr)-1):
            if arr[i+1] - arr[i] != dif:
                return False
        return True
        
class Solution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        arr = sorted(arr)
        dif = arr[1] - arr[0]
        array = [arr[i+1] - arr[i] for i in range(len(arr)-1)]
        return len(set(array)) == 1
      
class Solution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        arr = sorted(arr)
        i = arr[:len(arr)-1]
        j = arr[1:]
        array = [i - j for i, j in zip(i, j)]
        return len(set(array)) == 1
        

File: Leetcode/cpp/1502.cpp
class Solution {
public:
  bool canMakeArithmeticProgression(vector<int>& arr) {
    int min_value = *min_element(arr.begin(), arr.end());
    int max_value = *max_element(arr.begin(), arr.end());
    int n = arr.size();
    if ((max_value - min_value) == 0) {
      return true;
    } else if ((max_value - min_value) % (n - 1) != 0) {
      return false;
    }
    int diff = (max_value - min_value) / (n - 1);
    unordered_set<int> number_set;
    for (auto num : arr) {
        if ((num - min_value) % diff != 0) {
          return false;
        }
        number_set.insert(num);
    }
    return number_set.size() == n;
  }
};


class Solution {
public:
  bool canMakeArithmeticProgression(vector<int>& arr) {
    std::sort(arr.begin(), arr.end());
    int d = arr[1] - arr[0];
    for (int i = 2; i < arr.size(); i++) {
        if (arr[i] - arr[i - 1] != d) {
            return false;
        }
    }
    return true;
  }
};
